costmap_2d/include/costmap_2d/footprint.h

```diff
@@ -40,18 +40,106 @@
 
 #include <ros/ros.h>
 #include <geometry_msgs/Polygon.h>
+#include <geometry_msgs/PolygonStamped.h>
 #include <geometry_msgs/Point.h>
 #include <geometry_msgs/Point32.h>
 
 namespace costmap_2d
 {
 
+/**
+ * @brief Calculate the extreme distances for the footprint
+ * 
+ * @param footprint The footprint to examine
+ * @param min_dist Output parameter of the minimum distance
+ * @param max_dist Output parameter of the maximum distance 
+ */
 void calculateMinAndMaxDistances(const std::vector<geometry_msgs::Point>& footprint, double& min_dist, double& max_dist);
+
+/**
+ * @brief Convert Point32 to Point
+ */
 geometry_msgs::Point              toPoint      (geometry_msgs::Point32 pt);
+
+/**
+ * @brief Convert Point to Point32
+ */
 geometry_msgs::Point32            toPoint32    (geometry_msgs::Point   pt);
+
+/**
+ * @brief Convert vector of Points to Polygon msg
+ */
 geometry_msgs::Polygon            toPolygon    (std::vector<geometry_msgs::Point> pts);
+
+/**
+ * @brief Convert Polygon msg to vector of Points. 
+ */
 std::vector<geometry_msgs::Point> toPointVector(geometry_msgs::Polygon polygon);
 
+/**
+ * @brief  Given a pose and base footprint, build the oriented footprint of the robot (list of Points)
+ * @param  x The x position of the robot
+ * @param  y The y position of the robot
+ * @param  theta The orientation of the robot
+ * @param  footprint_spec Basic shape of the footprint
+ * @param  oriented_footprint Will be filled with the points in the oriented footprint of the robot
+*/
+void transformFootprint(double x, double y, double theta, const std::vector<geometry_msgs::Point>& footprint_spec, std::vector<geometry_msgs::Point>& oriented_footprint);
+
+/**
+ * @brief  Given a pose and base footprint, build the oriented footprint of the robot (PolygonStamped)
+ * @param  x The x position of the robot
+ * @param  y The y position of the robot
+ * @param  theta The orientation of the robot
+ * @param  footprint_spec Basic shape of the footprint
+ * @param  oriented_footprint Will be filled with the points in the oriented footprint of the robot
+*/
+void transformFootprint(double x, double y, double theta, const std::vector<geometry_msgs::Point>& footprint_spec, geometry_msgs::PolygonStamped & oriented_footprint);
+
+/**
+ * @brief Adds the specified amount of padding to the footprint (in place)
+ */
+void padFootprint( std::vector<geometry_msgs::Point>& footprint, double padding );
+
+/**
+ * @brief Create a circular footprint from a given radius
+ */
+std::vector<geometry_msgs::Point> makeFootprintFromRadius( double radius );
+
+/** 
+ * @brief Make the footprint from the given string.
+ *
+ * Format should be bracketed array of arrays of floats, like so: [[1.0, 2.2], [3.3, 4.2], ...] 
+ * 
+ * If there's an error, the error_flag will be set to true. 
+ */
+std::vector<geometry_msgs::Point> makeFootprintFromString( const std::string& footprint_string, bool* error_flag );
```

Comment:
this seems an odd API -- and the usage below could have unintentened consequences. I would suggest that this be changed so that the vector of points is passed in by reference, the function returns bool, and then when used you can say if(makeFootprintFromString(string, footprint)) { setUnpaddedRobotFootprint(footprint) } else { print error message }